Custom component extension type & extension preferences (#10443)
owncloud/web: docs/extension-system/extension-types/custom-components.md

---
title: 'Custom component extensions'
date: 2024-02-14T00:00:00+00:00
weight: 60
geekdocRepo: https://github.com/owncloud/web
geekdocEditPath: edit/master/docs/extension-system/extension-types
geekdocFilePath: custom-components.md
geekdocCollapseSection: true
---

{{< toc >}}

## Extension Type CustomComponent

CustomComponent extensions need to define one or multiple `extensionPointId`s as render target. A `CustomComponentTarget` component for this very 
extension point needs to be mounted in the current view.

### Configuration

To define a custom component extension, you implement the `CustomComponentExtension` interface.

> Here's how it looks like:

Reviewer: ```suggestion
Here's what it looks like:
```

```typescript
interface CustomComponentExtension {
  id: string,
  type: 'customComponent',
  extensionPointIds: string[],
  content: Slot
}
```

For `id`, `type`, and `extensionPointIds`, please see [extension base section]({{< ref "../_index.md#extension-base-configuration" >}}) in the top level docs.

> The `content` object configures the actual custom component to render in the target extension point.

Reviewer: ```suggestion
The `content` property specifies a render function for the target extension point.
```

### Example

> A simple example for a custom component extension would be the `NyanCat` progress bar component, being

Reviewer: ```suggestion
A simple example for a custom component extension could be a `NyanCat` progress bar component, being
```

targeted at the `global-progress-bar` extension point as render target.

```typescript
({
    id: 'com.github.owncloud.web.app.progress-bars.nyan-cat',
    type: 'customComponent',
    extensionPointIds: ['app.runtime.global-progress-bar'],
    content: (slots) => [h(NyanCat, slots)],
    userPreference: {
      optionLabel: $gettext('Nyan Cat progress bar')
    }

> })

Reviewer: What are the outer parentheses good for?

Author: For syntax highlighting not complaining about invalid typescript. I now assigned the object to a `const`, that works as well... is that ok from your POV?

Reviewer: Fine either way
